File: javascript/Intermediate-Algorithm-Scripting/spinalTapCase.js

/*

Convert a string to spinal case. Spinal case is all-lowercase-words-joined-by-dashes.

*/

function spinalCase(str) {
    
    function lastChar(str) {
        return str.charAt(str.length-1);
    }

    let ret = "";
    for(let i=0; i<str.length; i++) {
        let char = str.charAt(i);
        if(char=='_' || char=='-' || char==' ')
            ret += ((lastChar(ret) != '-' && i>0) ? '-' : '');
        else if (char.toUpperCase() === char)
            ret += (((lastChar(ret) != '-' && i>0) ? '-' : '') + char.toLowerCase());
        else
            ret += char;
    }
    return ret;
}

spinalCase('This Is Spinal Tap');

File: python/Intermediate-Algorithm-Scripting/spinalTapCase.py

'''

Convert a string to spinal case. Spinal case is all-lowercase-words-joined-by-dashes.

'''

def spinalCase(inp_str):
    ret = ''
    for ix, char in enumerate(inp_str.strip()):
        if char in ('_', '-', ' '):
            if ix>0 and ret[-1] != '-':
                ret += '-'
        elif char.isupper():
            if ix>0 and ret[-1] != '-':
                ret += '-'
            ret += char.lower()
        else:
            ret += char
    return ret

spinalCase('This Is Spinal Tap')